toponetx/classes/path.py

```diff
@@ -0,0 +1,227 @@
+    attr: keyword arguments, optional
+        Additional attributes to be associated with the elementary p-path.
+
+    Notes
+    -----
+    - An elementary p-path is defined as an ordered sequence of nodes (n1, ..., np) if we reserve the sequence order. Thus, for this case, (n1, ..., np) is different
+    from (np, ..., n1). If we do not reserve the sequence order, then (n1, ..., np) is the same as (np, ..., n1). For this case, the first index must be smaller
+    than the last index so that we can discard the other duplicate elementary p-path. As elementary p-paths may contain characters and numbers at the same time, we leverage
+    lexicographical order to compare two indices.
+    - Similarly to simplex, in order to find the boundary of an elementary p-path, we remove one node at a time from the elementary p-path. However, unlike simplex
+    where order does not matter, the process of omitting nodes from an elementary p-path may result some non-existing paths. For instance, if we have an elementary
+    p-path (1, 2, 3) and there is no path (1, 3), then applying boundary operation on the elementary p-path results in [(2,3), (1,3), (1,2)]. In order to avoid this,
+    we can provide a list of allowed boundaries. If the boundary of an elementary p-path is not in the list of allowed boundaries,
+    then it will not be included in the boundary of the elementary p-path.
+    - When an elementary p-path is created and allowed paths are not specified, its boundary is automatically created by iteratively removing one node at a time
+    from the elementary p-path, which is identical to a simplex.
+
+    References
+    ----------
+    .. [1] Truong and Chin.
+        Generalizing Topological Graph Neural Networks with Paths.
+        https://arxiv.org/abs/2308.06838.pdf
+    .. [2] Grigor'yan, Lin, Muranov, and Yau.
+        Homologies of path complexes and digraphs.
+        https://arxiv.org/pdf/1207.2834.pdf
+
+    Examples
+    --------
+    >>> path1 = Path((1, 2, 3))
+    >>> list(path1.boundary)
+    []
+    >>> path2 = Path((1, 2, 3), construct_boundaries=True)
+    >>> list(path2.boundary)
+    [(2, 3), (1, 3), (1, 2)]
+    >>> path3 = Path((1, 2, 3), construct_boundaries=True, allowed_paths=[(1, 2), (2, 3)])
+    >>> list(path3.boundary)
+    [(2, 3), (1, 2)]
+    >>> path4 = Path((1, 2, 3), construct_boundaries=True, allowed_paths=[(1, 3), (2, 3)])
+    >>> list(path4.boundary)
+    [(2, 3), (1, 3)]
+    """
+
+    def __init__(
+        self,
+        elements: Sequence[Hashable],
+        name: str = "",
+        construct_boundaries: bool = False,
+        reserve_sequence_order: bool = False,
+        allowed_paths: Iterable[tuple[Hashable]] | None = None,
+        **attr,
+    ) -> None:
+        self.__check_inputs(elements, reserve_sequence_order)
+        if isinstance(elements, int) or isinstance(elements, str):
```

Comment:
This cannot happen according to the type definition of `elements`. Either support this explicitly or remove this (I would prefer the later).

Reply:
I already have some test cases covering this, so I still keep and explicitly support it in function signature. `__check_inputs()` also supports this from the beginning. My intuition is that users may prefer to explitcitly call `Path(3)` instead of `Path([3])` when instantiating a 0-path.